x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DownsamplePrepareLifeCycleStateStep.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.ilm;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.LifecycleExecutionState;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.search.aggregations.bucket.histogram.DateHistogramInterval;
+
+import static org.elasticsearch.xpack.core.ilm.DownsampleAction.DOWNSAMPLED_INDEX_PREFIX;
+
+public class DownsamplePrepareLifeCycleStateStep extends ClusterStateActionStep {
+
+    private static final Logger LOGGER = LogManager.getLogger(DownsamplePrepareLifeCycleStateStep.class);
+    public static final String NAME = "generate-downsampled-index-name";
+    private final DateHistogramInterval fixedInterval;
+
+    public DownsamplePrepareLifeCycleStateStep(StepKey key, StepKey nextStepKey, DateHistogramInterval fixedInterval) {
+        super(key, nextStepKey);
+        this.fixedInterval = fixedInterval;
+    }
+
+    @Override
+    public ClusterState performAction(Index index, ClusterState clusterState) {
+        IndexMetadata indexMetadata = clusterState.metadata().index(index);
+        if (indexMetadata == null) {
+            // Index must have been since deleted, ignore it
+            LOGGER.debug("[{}] lifecycle action for index [{}] executed but index no longer exists", getKey().action(), index.getName());
+            return clusterState;
+        }
+
+        LifecycleExecutionState lifecycleState = indexMetadata.getLifecycleExecutionState();
+
+        LifecycleExecutionState.Builder newLifecycleState = LifecycleExecutionState.builder(lifecycleState);
+        final String downsampleIndexName = generateDownsampleIndexName(indexMetadata, fixedInterval);
+        newLifecycleState.setDownsampleIndexName(downsampleIndexName);
+
+        return LifecycleExecutionStateUtils.newClusterStateWithLifecycleState(
+            clusterState,
+            indexMetadata.getIndex(),
+            newLifecycleState.build()
+        );
+    }
+
+    @Override
+    public boolean isRetryable() {
+        return true;
+    }
+
+    static String generateDownsampleIndexName(IndexMetadata sourceIndexMetadata, DateHistogramInterval fixedInterval) {
```

Comment:
Hm, I'm not sure we need this persisted in the lifecycle execution state. 

Could we not derive this exact logic each time in `DownsampleStep`?

Reply:
Maybe I'm not understanding something correctly. If there is an easier way to do this then that would be great. But the issue here is that also the `downsampleAllocatedStep`, `copyExecutionStateStep`, `copyLifecycleSettingsStep`, `replaceDataStreamBackingIndex` and `swapAliasesAndDeleteSourceIndexStep` depend on the target name being generated. All step instances in the downsample action are shared, right? So it felt that the easiest way was to reuse the `LfecycleState#downsampleIndexName()` again.

Reply:
++ this is what I was missing

So in theory we wouldn't need the persistence in the lifecycle state however the other steps are used across other actions and their refactoring would entail quite a bit of effort (as other actions do rely on index names coming from the lifecycle execution state as they are randomly generated)